File: go/archive-metrics.go
// Copyright 2024 Coralogix Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cxsdk

import (
	"context"

	archiveMetrics "github.com/coralogix/coralogix-management-sdk/go/internal/coralogix/metrics"

	"google.golang.org/protobuf/types/known/emptypb"
)

// ConfigureTenantRequest is a request to configure the tenant.
type ConfigureTenantRequest = archiveMetrics.ConfigureTenantRequest

// UpdateTenantRequest is a request to update the tenant.
type UpdateTenantRequest = archiveMetrics.UpdateRequest

// ValidateBucketRequest is a request to validate the bucket.
type ValidateBucketRequest = archiveMetrics.ValidateBucketRequest

// GetTenantConfigRequest is a request to get the tenant configuration.
type GetTenantConfigRequest = archiveMetrics.GetTenantConfigRequest

// ConfigureTenantRequestS3 is a request to configure the tenant with S3.
type ConfigureTenantRequestS3 = archiveMetrics.ConfigureTenantRequest_S3

// ConfigureTenantRequestIbm is a request to configure the tenant with Ibm.
type ConfigureTenantRequestIbm = archiveMetrics.ConfigureTenantRequest_Ibm

// ArchiveIbmConfigV2 is an Ibm configuration for the archive.
type ArchiveIbmConfigV2 = archiveMetrics.IbmConfigV2

// ArchiveS3Config is an S3 configuration for the archive.
type ArchiveS3Config = archiveMetrics.S3Config

// ValidateBucketRequestS3 is a request to validate the S3 bucket.
type ValidateBucketRequestS3 = archiveMetrics.ValidateBucketRequest_S3

// ValidateBucketRequestIbm is a request to validate the IBM storage.
type ValidateBucketRequestIbm = archiveMetrics.ValidateBucketRequest_Ibm

// UpdateRequestS3 is a type to update the S3 bucket.
type UpdateRequestS3 = archiveMetrics.UpdateRequest_S3

// UpdateRequestIbm is a type to update the IBM storage.
type UpdateRequestIbm = archiveMetrics.UpdateRequest_Ibm

// TenantConfigV2S3 is a type to view the S3 bucket config.
type TenantConfigV2S3 = archiveMetrics.TenantConfigV2_S3

// TenantConfigV2Ibm is a type to view the IBM storage config.
type TenantConfigV2Ibm = archiveMetrics.TenantConfigV2_Ibm

// TenantConfigV2 is a type to configure the tenant.
type TenantConfigV2 = archiveMetrics.TenantConfigV2

// RetentionPolicyRequest is a request to set the retention policy.
type RetentionPolicyRequest = archiveMetrics.RetentionPolicyRequest

const archiveMetricsFeatureGroupID = "metrics"

// RPC names.
const (
	ArchiveMetricsConfigureTenantRPC = archiveMetrics.MetricsConfiguratorPublicService_ConfigureTenant_FullMethodName
	ArchiveMetricsUpdateRPC          = archiveMetrics.MetricsConfiguratorPublicService_Update_FullMethodName
	ArchiveMetricsValidateBucketRPC  = archiveMetrics.MetricsConfiguratorPublicService_ValidateBucket_FullMethodName
	ArchiveMetricsGetTenantConfigRPC = archiveMetrics.MetricsConfiguratorPublicService_GetTenantConfig_FullMethodName
	ArchiveMetricsEnableArchiveRPC   = archiveMetrics.MetricsConfiguratorPublicService_EnableArchive_FullMethodName
	ArchiveMetricsDisableArchiveRPC  = archiveMetrics.MetricsConfiguratorPublicService_DisableArchive_FullMethodName
)

// ArchiveMetricsClient is a client for the Coralogix Archive Metrics API.
type ArchiveMetricsClient struct {
	callPropertiesCreator *CallPropertiesCreator
}

// Update updates the archive metrics configuration.
func (c ArchiveMetricsClient) Update(ctx context.Context, req *archiveMetrics.UpdateRequest) (*emptypb.Empty, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveMetrics.NewMetricsConfiguratorPublicServiceClient(conn)

	response, err := client.Update(callProperties.Ctx, req, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveMetricsUpdateRPC, archiveMetricsFeatureGroupID)
	}
	return response, nil
}

// Get gets the archive metrics configuration.
func (c ArchiveMetricsClient) Get(ctx context.Context) (*archiveMetrics.GetTenantConfigResponseV2, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveMetrics.NewMetricsConfiguratorPublicServiceClient(conn)

	response, err := client.GetTenantConfig(callProperties.Ctx, &emptypb.Empty{}, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveMetricsGetTenantConfigRPC, archiveMetricsFeatureGroupID)
	}
	return response, nil
}

// ConfigureTenant configures the archive metrics bucket.
func (c ArchiveMetricsClient) ConfigureTenant(ctx context.Context, req *archiveMetrics.ConfigureTenantRequest) (*emptypb.Empty, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveMetrics.NewMetricsConfiguratorPublicServiceClient(conn)

	response, err := client.ConfigureTenant(callProperties.Ctx, req, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveMetricsConfigureTenantRPC, archiveMetricsFeatureGroupID)
	}
	return response, nil
}

// ValidateTarget validates the archive metrics bucket.
func (c ArchiveMetricsClient) ValidateTarget(ctx context.Context, req *archiveMetrics.ValidateBucketRequest) (*emptypb.Empty, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveMetrics.NewMetricsConfiguratorPublicServiceClient(conn)

	response, err := client.ValidateBucket(callProperties.Ctx, req, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveMetricsValidateBucketRPC, archiveMetricsFeatureGroupID)
	}
	return response, nil
}

// Enable enables the metrics archive.
func (c ArchiveMetricsClient) Enable(ctx context.Context) (*emptypb.Empty, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveMetrics.NewMetricsConfiguratorPublicServiceClient(conn)

	response, err := client.EnableArchive(callProperties.Ctx, &emptypb.Empty{}, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveMetricsEnableArchiveRPC, archiveMetricsFeatureGroupID)
	}
	return response, nil
}

// Disable disables the metrics archive.
func (c ArchiveMetricsClient) Disable(ctx context.Context) (*emptypb.Empty, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveMetrics.NewMetricsConfiguratorPublicServiceClient(conn)

	response, err := client.DisableArchive(callProperties.Ctx, &emptypb.Empty{}, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveMetricsDisableArchiveRPC, archiveMetricsFeatureGroupID)
	}
	return response, nil
}

// NewArchiveMetricsClient creates a new archive metrics client.
func NewArchiveMetricsClient(c *CallPropertiesCreator) *ArchiveMetricsClient {
	return &ArchiveMetricsClient{callPropertiesCreator: c}
}

File: rust/cx-sdk/src/client/archive_metrics.rs
// Copyright 2024 Coralogix Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use cx_api::proto::com::coralogix::metrics::metrics_configurator::{
    metrics_configurator_public_service_client::MetricsConfiguratorPublicServiceClient,
    metrics_configurator_service_client::MetricsConfiguratorServiceClient,
    ConfigureTenantRequest,
    GetTenantConfigRequest,
    GetTenantConfigResponse,
    GetTenantConfigResponseV2,
    InternalUpdateRequest,
    ListHotStoreConfigsRequest,
    ListHotStoreConfigsResponse,
    ListTenantConfigsRequest,
    ListTenantConfigsResponse,
    MigrateTenantRequest,
    UpdateRequest,
    ValidateBucketRequest,
};
use std::str::FromStr;
use tokio::sync::Mutex;
use tonic::{
    metadata::MetadataMap,
    transport::{
        Channel,
        ClientTlsConfig,
        Endpoint,
    },
};

use crate::{
    auth::AuthContext,
    error::Result,
    metadata::CallProperties,
    util::make_request_with_metadata,
    CoralogixRegion,
};

pub use cx_api::proto::com::coralogix::metrics::metrics_configurator::{
    configure_tenant_request::StorageConfig,
    internal_update_request::StorageConfig as InternalStorageConfigUpdate,
    tenant_config::StorageConfig as InternalStorageConfig,
    tenant_config_v2::StorageConfig as StorageConfigView,
    update_request::StorageConfig as StorageConfigUpdate,
    validate_bucket_request::StorageConfig as StorageConfigValidation,
    RetentionPolicyRequest,
    S3Config,
};

/// The metrics archive API client.
/// Read more at [https://coralogix.com/docs/archive-s3-bucket-forever/]()
pub struct MetricsArchiveClient {
    metadata_map: MetadataMap,
    service_client: Mutex<MetricsConfiguratorPublicServiceClient<Channel>>,
}

impl MetricsArchiveClient {
    /// Creates a new client for the Metrics Archive API.
    ///
    /// # Arguments
    /// * `auth_context` - The  to use for authentication.
    /// * `region` - The region to connect to.
    pub fn new(region: CoralogixRegion, auth_context: AuthContext) -> Result<Self> {
        let channel: Channel = Endpoint::from_str(&region.grpc_endpoint())?
            .tls_config(ClientTlsConfig::new().with_native_roots())?
            .connect_lazy();
        let request_metadata: CallProperties = (&auth_context.team_level_api_key).into();
        Ok(Self {
            metadata_map: request_metadata.to_metadata_map(),
            service_client: Mutex::new(MetricsConfiguratorPublicServiceClient::new(channel)),
        })
    }

    /// Configures the tenant.
    ///
    /// # Arguments
    /// * `retention_policy` - The retention policy to set.
    /// * `storage_config` - The storage configuration to set.
    pub async fn configure_tenant(
        &self,
        retention_policy: Option<RetentionPolicyRequest>,
        storage_config: StorageConfig,
    ) -> Result<()> {
        let request = make_request_with_metadata(
            ConfigureTenantRequest {
                retention_policy,
                storage_config: Some(storage_config),
            },
            &self.metadata_map,
        );
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .configure_tenant(request)
                .await
                .map(|_| ())
                .map_err(From::from)
        }
    }

    /// Updates the tenant configuration.
    ///
    /// # Arguments
    /// * `retention_days` - The retention days to set.
    /// * `storage_config` - The storage configuration to set.
    pub async fn update_tenant(
        &self,
        retention_days: u32,
        storage_config: StorageConfigUpdate,
    ) -> Result<()> {
        let request = make_request_with_metadata(
            UpdateRequest {
                retention_days: Some(retention_days),
                storage_config: Some(storage_config),
            },
            &self.metadata_map,
        );
        {
            let mut client = self.service_client.lock().await.clone();

            client.update(request).await.map(|_| ()).map_err(From::from)
        }
    }

    /// Validates a bucket configuration.
    ///
    /// # Arguments
    /// * `storage_config` - The storage configuration to validate.
    /// * `bucket_name` - The name of the bucket to validate.
    pub async fn validate_bucket(&self, storage_config: StorageConfigValidation) -> Result<()> {
        let request = make_request_with_metadata(
            ValidateBucketRequest {
                storage_config: Some(storage_config),
            },
            &self.metadata_map,
        );
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .validate_bucket(request)
                .await
                .map(|_| ())
                .map_err(From::from)
        }
    }

    /// Gets the tenant configuration.
    pub async fn get_tenant_config(&self) -> Result<GetTenantConfigResponseV2> {
        let request = make_request_with_metadata((), &self.metadata_map);
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .get_tenant_config(request)
                .await
                .map(|response| response.into_inner())
                .map_err(From::from)
        }
    }

    /// Enables the archive.
    pub async fn enable_archive(&self) -> Result<()> {
        let request = make_request_with_metadata((), &self.metadata_map);
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .enable_archive(request)
                .await
                .map(|_| ())
                .map_err(From::from)
        }
    }

    /// Disables the archive.
    pub async fn disable_archive(&self) -> Result<()> {
        let request = make_request_with_metadata((), &self.metadata_map);
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .disable_archive(request)
                .await
                .map(|_| ())
                .map_err(From::from)
        }
    }
}

/// A service client for the internal metrics archive API. It's only for Coralogix internal use.
pub struct MetricsArchiveInternalClient {
    metadata_map: MetadataMap,
    service_client: Mutex<MetricsConfiguratorServiceClient<Channel>>,
}

impl MetricsArchiveInternalClient {
    /// Creates a new client for the internal Metrics Archive API.
    ///
    /// # Arguments
    /// * `auth_context` - The  to use for authentication.
    /// * `region` - The region to connect to.
    pub fn new(region: CoralogixRegion, auth_context: AuthContext) -> Result<Self> {
        let channel: Channel = Endpoint::from_str(&region.grpc_endpoint())?
            .tls_config(ClientTlsConfig::new().with_native_roots())?
            .connect_lazy();
        let request_metadata: CallProperties = (&auth_context.team_level_api_key).into();
        Ok(Self {
            metadata_map: request_metadata.to_metadata_map(),
            service_client: Mutex::new(MetricsConfiguratorServiceClient::new(channel)),
        })
    }

    /// Gets the tenant configuration.
    ///
    /// # Arguments
    /// * `tenant_id` - The ID of the tenant to get the configuration for.
    pub async fn get_tenant_config(&self, tenant_id: u32) -> Result<GetTenantConfigResponse> {
        let request =
            make_request_with_metadata(GetTenantConfigRequest { tenant_id }, &self.metadata_map);
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .get_tenant_config(request)
                .await
                .map(|response| response.into_inner())
                .map_err(From::from)
        }
    }

    /// Lists the tenant configurations.
    pub async fn list_tenant_configs(&self) -> Result<ListTenantConfigsResponse> {
        let request = make_request_with_metadata(ListTenantConfigsRequest {}, &self.metadata_map);
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .list_tenant_configs(request)
                .await
                .map(|response| response.into_inner())
                .map_err(From::from)
        }
    }

    /// Lists the hot store configurations.
    pub async fn list_hot_store_configs(&self) -> Result<ListHotStoreConfigsResponse> {
        let request = make_request_with_metadata(ListHotStoreConfigsRequest {}, &self.metadata_map);
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .list_host_store_configs(request)
                .await
                .map(|r| r.into_inner())
                .map_err(From::from)
        }
    }

    /// Migrates a tenant.
    ///
    /// # Arguments
    /// * `tenant_id` - The ID of the tenant to migrate.
    pub async fn migrate_tenant(&self, tenant_id: u32) -> Result<()> {
        let request =
            make_request_with_metadata(MigrateTenantRequest { tenant_id }, &self.metadata_map);
        {
            let mut client = self.service_client.lock().await.clone();

            client
                .migrate_tenant(request)
                .await
                .map(|_| ())
                .map_err(From::from)
        }
    }

    /// Updates the tenant configuration.
    ///
    /// # Arguments
    /// * `tenant_id` - The ID of the tenant to update.
    pub async fn update(
        &self,
        tenant_id: u32,
        retention_days: u32,
        storage_config: InternalStorageConfigUpdate,
    ) -> Result<()> {
        let request = make_request_with_metadata(
            InternalUpdateRequest {
                retention_days: Some(retention_days),
                storage_config: Some(storage_config),
                tenant_id,
            },
            &self.metadata_map,
        );
        {
            let mut client = self.service_client.lock().await.clone();

            client.update(request).await.map(|_| ()).map_err(From::from)
        }
    }
}
